File: insertion_sort/go/insertion_sort.go
package insertionsort

func InsertionSort(A []int) []int {
	n := len(A)
	for i := 1; i < n; i++ {
		key := A[i]
		var k int
		for j := i - 1; (j >= 0) && (A[j] > key); j-- {
			A[j+1] = A[j]
			k = j
		}
		A[k] = key
	}
	return A
}

File: insertion_sort/python/insertion_sort.py


def insertion_sort(A: list[int]):
    """
    >>> A = [1,45,2,3,4]
    >>> insertion_sort(A)
    [1, 2, 3, 4, 45]
    """
    n = len(A)
    for i in range(n):
        key = A[i]
        j = i -1
        while A[j] > key and j > 0:
            A[j+1] = A[j]
            j = j - 1
        A[j+1] = key
    return A
if __name__ == "__main__":
    import doctest
    doctest.testmod() #test the whole module.